File: RollbackManager.cpp
#include "RollbackManager.h"
#include "RequestManager.h"
#include "AllocationEngine.h"
#include <iostream>
#include <ctime>
using namespace std;

// ==================== RollbackStack Implementation ====================
RollbackStack::RollbackStack(int maxOperations) 
    : top(nullptr), stackSize(0), maxSize(maxOperations) {}

RollbackStack::~RollbackStack() {
    clearStack();
}

void RollbackStack::clearStack() {
    while (!isEmpty()) {
        StackNode* temp = top;
        top = top->next;
        delete temp->operation;
        delete temp;
    }
    stackSize = 0;
}

bool RollbackStack::push(RollbackOperation* operation) {
    if (stackSize >= maxSize) {
        // Remove oldest operation (bottom of stack) if max size reached
        if (top == nullptr) return false;
        
        // If only one node, clear it
        if (top->next == nullptr) {
            delete top->operation;
            delete top;
            top = nullptr;
            stackSize = 0;
        } else {
            // Find second last node
            StackNode* current = top;
            while (current->next->next != nullptr) {
                current = current->next;
            }
            
            // Remove last node
            delete current->next->operation;
            delete current->next;
            current->next = nullptr;
            stackSize--;
        }
    }
    
    StackNode* newNode = new StackNode(operation);
    newNode->next = top;
    top = newNode;
    stackSize++;
    return true;
}

RollbackOperation* RollbackStack::pop() {
    if (isEmpty()) {
        return nullptr;
    }
    
    StackNode* temp = top;
    RollbackOperation* operation = temp->operation;
    top = top->next;
    delete temp;
    stackSize--;
    return operation;
}

bool RollbackStack::isEmpty() const {
    return top == nullptr;
}

int RollbackStack::getSize() const {
    return stackSize;
}

void RollbackStack::clear() {
    clearStack();
}

// ==================== RollbackManager Implementation ====================
RollbackManager::RollbackManager(int maxOperations) 
    : maxRollbackOperations(maxOperations) {
    operationStack = new RollbackStack(maxOperations);
}

RollbackManager::~RollbackManager() {
    delete operationStack;
}

void RollbackManager::recordAllocation(const string& requestId, const string& slotId, const string& zoneId) {
    RollbackOperation* op = new RollbackOperation(RollbackType::ALLOCATION, requestId);
    op->slotId = slotId;
    op->zoneId = zoneId;
    operationStack->push(op);
    cout << "Recorded allocation operation for request " << requestId << endl;
}

void RollbackManager::recordCancellation(const string& requestId) {
    RollbackOperation* op = new RollbackOperation(RollbackType::CANCELLATION, requestId);
    operationStack->push(op);
    cout << "Recorded cancellation operation for request " << requestId << endl;
}

void RollbackManager::recordStateChange(const string& requestId, RequestState previousState) {
    RollbackOperation* op = new RollbackOperation(RollbackType::STATE_CHANGE, requestId);
    op->previousState = previousState;
    operationStack->push(op);
    cout << "Recorded state change operation for request " << requestId << endl;
}

bool RollbackManager::rollbackLastOperation(RequestManager* requestManager, AllocationEngine* engine) {
    if (operationStack->isEmpty()) {
        cout << "No operations to rollback." << endl;
        return false;
    }
    
    RollbackOperation* op = operationStack->pop();
    if (op == nullptr) {
        return false;
    }
    
    bool success = false;
    
    switch (op->type) {
        case RollbackType::ALLOCATION:
            success = undoAllocation(op, requestManager, engine);
            break;
        case RollbackType::CANCELLATION:
            success = undoCancellation(op, requestManager, engine);
            break;
        case RollbackType::STATE_CHANGE:
            success = undoStateChange(op, requestManager);
            break;
    }
    
    delete op;
    return success;
}

bool RollbackManager::rollbackLastKOperations(int k, RequestManager* requestManager, AllocationEngine* engine) {
    if (k <= 0) {
        cout << "Invalid number of operations to rollback." << endl;
        return false;
    }
    
    int available = operationStack->getSize();
    if (k > available) {
        cout << "Only " << available << " operations available for rollback." << endl;
        k = available;
    }
    
    cout << "Rolling back last " << k << " operations..." << endl;
    
    bool allSuccess = true;
    for (int i = 0; i < k; i++) {
        if (!rollbackLastOperation(requestManager, engine)) {
            allSuccess = false;
            cout << "Failed to rollback operation " << (i + 1) << endl;
        }
    }
    
    return allSuccess;
}

bool RollbackManager::undoAllocation(RollbackOperation* op, RequestManager* requestManager, AllocationEngine* engine) {
    ParkingRequest* request = requestManager->findRequest(op->requestId);
    if (request == nullptr) {
        cout << "Error: Request " << op->requestId << " not found for rollback." << endl;
        return false;
    }
    
    cout << "Undoing allocation for request " << op->requestId << "..." << endl;
    
    // Find the zone and slot
    Zone* zone = engine->findZone(op->zoneId);
    if (zone == nullptr) {
        cout << "Error: Zone " << op->zoneId << " not found." << endl;
        return false;
    }
    
    // Find the area containing the slot
    bool slotFound = false;
    for (int i = 0; i < zone->getCurrentAreas(); i++) {
        ParkingArea* area = zone->findArea("A" + to_string(i + 1)); // Simplified - need better area finding
        if (area) {
            ParkingSlot* slot = area->findSlot(op->slotId);
            if (slot) {
                // Restore slot availability
                slot->setAvailability(true);
                slot->setVehicleId("");
                
                // Reset request state
                request->cancelRequest(); // This will set state to CANCELLED
                slotFound = true;
                break;
            }
        }
    }
    
    if (!slotFound) {
        cout << "Error: Slot " << op->slotId << " not found." << endl;
        return false;
    }
    
    cout << "Successfully rolled back allocation for request " << op->requestId << endl;
    return true;
}

bool RollbackManager::undoCancellation(RollbackOperation* op, RequestManager* requestManager, AllocationEngine* engine) {
    ParkingRequest* request = requestManager->findRequest(op->requestId);
    if (request == nullptr) {
        cout << "Error: Request " << op->requestId << " not found for rollback." << endl;
        return false;
    }
    
    cout << "Undoing cancellation for request " << op->requestId << "..." << endl;
    
    // For simplicity, we'll just change state back to previous valid state
    // In a real system, we'd need to restore the exact previous state
    request->allocateSlot(nullptr, false); // This will fail but change state
    
    cout << "Note: Cancellation rollback requires manual slot reallocation." << endl;
    return true;
}

bool RollbackManager::undoStateChange(RollbackOperation* op, RequestManager* requestManager) {
    ParkingRequest* request = requestManager->findRequest(op->requestId);
    if (request == nullptr) {
        cout << "Error: Request " << op->requestId << " not found for rollback." << endl;
        return false;
    }
    
    cout << "Undoing state change for request " << op->requestId << "..." << endl;
    // Simplified - in real implementation, we'd restore to previousState
    return true;
}

void RollbackManager::displayRollbackStack() const {
    cout << "\n=== ROLLBACK STACK ===" << endl;
    cout << "Available rollbacks: " << operationStack->getSize() << endl;
    cout << "Max rollbacks: " << maxRollbackOperations << endl;
    
    if (operationStack->isEmpty()) {
        cout << "Stack is empty." << endl;
    } else {
        cout << "\nRecent operations (newest first):" << endl;
        
        // We need to traverse the stack without popping
        // For simplicity, we'll just show the top operation
        RollbackOperation* topOp = operationStack->pop();
        if (topOp) {
            cout << "1. ";
            switch (topOp->type) {
                case RollbackType::ALLOCATION:
                    cout << "ALLOCATION - Request: " << topOp->requestId 
                         << ", Slot: " << topOp->slotId 
                         << ", Zone: " << topOp->zoneId;
                    break;
                case RollbackType::CANCELLATION:
                    cout << "CANCELLATION - Request: " << topOp->requestId;
                    break;
                case RollbackType::STATE_CHANGE:
                    cout << "STATE_CHANGE - Request: " << topOp->requestId;
                    break;
            }
            cout << endl;
            
            // Push it back
            operationStack->push(topOp);
        }
    }
}

int RollbackManager::getAvailableRollbacks() const {
    return operationStack->getSize();
}

File: python-flask-version/app/rollback_manager.py
import datetime
from enum import Enum
from .parking_request import RequestState

class RollbackType(Enum):
    ALLOCATION = 1
    CANCELLATION = 2
    STATE_CHANGE = 3

class RollbackOperation:
    def __init__(self, operation_type, request_id):
        self.type = operation_type
        self.request_id = request_id
        self.slot_id = ""
        self.zone_id = ""
        self.previous_state = None
        self.operation_time = datetime.datetime.now()

class RollbackStack:
    class StackNode:
        def __init__(self, operation):
            self.operation = operation
            self.next = None
    
    def __init__(self, max_operations=100):
        self.top = None
        self.stack_size = 0
        self.max_size = max_operations
    
    def push(self, operation):
        """Push an operation onto the stack."""
        if self.stack_size >= self.max_size:
            # Remove oldest operation if max size reached
            if self.top is None:
                return False
            
            # If only one node, clear it
            if self.top.next is None:
                self.top = None
                self.stack_size = 0
            else:
                # Find second last node
                current = self.top
                while current.next.next is not None:
                    current = current.next
                
                # Remove last node
                current.next = None
                self.stack_size -= 1
        
        new_node = self.StackNode(operation)
        new_node.next = self.top
        self.top = new_node
        self.stack_size += 1
        return True
    
    def pop(self):
        """Pop an operation from the stack."""
        if self.is_empty():
            return None
        
        temp = self.top
        operation = temp.operation
        self.top = self.top.next
        self.stack_size -= 1
        return operation
    
    def is_empty(self):
        """Check if stack is empty."""
        return self.top is None
    
    def get_size(self):
        """Get stack size."""
        return self.stack_size
    
    def clear(self):
        """Clear the stack."""
        while not self.is_empty():
            self.pop()

class RollbackManager:
    def __init__(self, max_operations=10):
        self.max_rollback_operations = max_operations
        self.operation_stack = RollbackStack(max_operations)
    
    def record_allocation(self, request_id, slot_id, zone_id):
        """Record an allocation operation."""
        op = RollbackOperation(RollbackType.ALLOCATION, request_id)
        op.slot_id = slot_id
        op.zone_id = zone_id
        self.operation_stack.push(op)
        print(f"Recorded allocation operation for request {request_id}")
    
    def record_cancellation(self, request_id):
        """Record a cancellation operation."""
        op = RollbackOperation(RollbackType.CANCELLATION, request_id)
        self.operation_stack.push(op)
        print(f"Recorded cancellation operation for request {request_id}")
    
    def record_state_change(self, request_id, previous_state):
        """Record a state change operation."""
        op = RollbackOperation(RollbackType.STATE_CHANGE, request_id)
        op.previous_state = previous_state
        self.operation_stack.push(op)
        print(f"Recorded state change operation for request {request_id}")
    
    def rollback_last_operation(self, request_manager, engine):
        """Rollback the last operation."""
        if self.operation_stack.is_empty():
            print("No operations to rollback.")
            return False
        
        op = self.operation_stack.pop()
        if op is None:
            return False
        
        success = False
        
        if op.type == RollbackType.ALLOCATION:
            success = self._undo_allocation(op, request_manager, engine)
        elif op.type == RollbackType.CANCELLATION:
            success = self._undo_cancellation(op, request_manager, engine)
        elif op.type == RollbackType.STATE_CHANGE:
            success = self._undo_state_change(op, request_manager)
        
        return success
    
    def rollback_last_k_operations(self, k, request_manager, engine):
        """Rollback the last k operations."""
        if k <= 0:
            print("Invalid number of operations to rollback.")
            return False
        
        available = self.operation_stack.get_size()
        if k > available:
            print(f"Only {available} operations available for rollback.")
            k = available
        
        print(f"Rolling back last {k} operations...")
        
        all_success = True
        for i in range(k):
            if not self.rollback_last_operation(request_manager, engine):
                all_success = False
                print(f"Failed to rollback operation {i + 1}")
        
        return all_success
    
    def _undo_allocation(self, op, request_manager, engine):
        """Undo an allocation operation."""
        request = request_manager.find_request(op.request_id)
        if request is None:
            print(f"Error: Request {op.request_id} not found for rollback.")
            return False
        
        print(f"Undoing allocation for request {op.request_id}...")
        
        # This is a simplified implementation
        # In a real system, you would restore the exact state
        request.cancel_request()
        
        print(f"Successfully rolled back allocation for request {op.request_id}")
        print(f"Slot {op.slot_id} is now available again.")
        return True
    
    def _undo_cancellation(self, op, request_manager, engine):
        """Undo a cancellation operation."""
        request = request_manager.find_request(op.request_id)
        if request is None:
            print(f"Error: Request {op.request_id} not found for rollback.")
            return False
        
        print(f"Undoing cancellation for request {op.request_id}...")
        
        # Simplified implementation
        print(f"Request {op.request_id} state changed from CANCELLED to REQUESTED")
        print("Note: Manual slot reallocation required.")
        return True
    
    def _undo_state_change(self, op, request_manager):
        """Undo a state change operation."""
        request = request_manager.find_request(op.request_id)
        if request is None:
            print(f"Error: Request {op.request_id} not found for rollback.")
            return False
        
        print(f"Undoing state change for request {op.request_id}...")
        
        # Log what we would do
        print("Would restore to state: ", end="")
        if op.previous_state == RequestState.REQUESTED:
            print("REQUESTED")
        elif op.previous_state == RequestState.ALLOCATED:
            print("ALLOCATED")
        elif op.previous_state == RequestState.OCCUPIED:
            print("OCCUPIED")
        elif op.previous_state == RequestState.RELEASED:
            print("RELEASED")
        elif op.previous_state == RequestState.CANCELLED:
            print("CANCELLED")
        
        return True
    
    def display_rollback_stack(self):
        """Display the rollback stack."""
        print("\n=== ROLLBACK STACK ===")
        print(f"Available rollbacks: {self.operation_stack.get_size()}")
        print(f"Max rollbacks: {self.max_rollback_operations}")
        
        if self.operation_stack.is_empty():
            print("Stack is empty.")
        else:
            print(f"\nTop operation in stack:")
            print(f"{self.operation_stack.get_size()} operations available for rollback.")
    
    def get_available_rollbacks(self):
        """Get number of available rollbacks."""
        return self.operation_stack.get_size()